File: src/main/java/net/lavaclient/client/event/EventManager.java
package net.lavaclient.client.event;

import java.lang.reflect.Method;
import java.util.*;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * Manager for handling and dispatching events
 */
public class EventManager {
    /**
     * Map of event classes to registered listeners
     */
    private final Map<Class<? extends Event>, List<EventData>> registeredEvents = new HashMap<>();
    
    /**
     * Registers a listener object
     * @param listener The listener object
     */
    public void register(Object listener) {
        // Find all methods with EventTarget annotation
        for (Method method : listener.getClass().getDeclaredMethods()) {
            if (method.isAnnotationPresent(EventTarget.class) && method.getParameterCount() == 1) {
                // Get the event class
                Class<?> eventClass = method.getParameterTypes()[0];
                
                // Ensure it's an event
                if (Event.class.isAssignableFrom(eventClass)) {
                    // Get the event priority
                    EventTarget annotation = method.getAnnotation(EventTarget.class);
                    
                    // Make the method accessible
                    if (!method.isAccessible()) {
                        method.setAccessible(true);
                    }
                    
                    // Get or create the event list
                    @SuppressWarnings("unchecked")
                    Class<? extends Event> castedEventClass = (Class<? extends Event>) eventClass;
                    
                    List<EventData> eventList = registeredEvents.computeIfAbsent(castedEventClass, k -> new CopyOnWriteArrayList<>());
                    
                    // Add the event data
                    eventList.add(new EventData(listener, method, annotation.priority()));
                    
                    // Sort by priority
                    eventList.sort(Comparator.comparingInt(EventData::getPriority));
                }
            }
        }
    }
    
    /**
     * Unregisters a listener object
     * @param listener The listener object
     */
    public void unregister(Object listener) {
        // Remove all event data for this listener
        for (List<EventData> eventList : registeredEvents.values()) {
            eventList.removeIf(eventData -> eventData.getListener() == listener);
        }
        
        // Remove empty lists
        registeredEvents.entrySet().removeIf(entry -> entry.getValue().isEmpty());
    }
    
    /**
     * Posts an event
     * @param event The event
     */
    public void post(Event event) {
        // Get the event list
        List<EventData> eventList = registeredEvents.get(event.getClass());
        
        // If no listeners, return
        if (eventList == null) {
            return;
        }
        
        // Notify all listeners
        for (EventData data : eventList) {
            try {
                data.invoke(event);
            } catch (Exception e) {
                System.err.println("Error posting event: " + e.getMessage());
                e.printStackTrace();
            }
            
            // If the event is cancelled, stop processing
            if (event.isCancelled()) {
                break;
            }
        }
    }
    
    /**
     * Class to hold event data
     */
    private static class EventData {
        private final Object listener;
        private final Method method;
        private final int priority;
        
        /**
         * Constructor
         * @param listener Listener object
         * @param method Listener method
         * @param priority Event priority
         */
        public EventData(Object listener, Method method, int priority) {
            this.listener = listener;
            this.method = method;
            this.priority = priority;
        }
        
        /**
         * Gets the listener object
         * @return Listener object
         */
        public Object getListener() {
            return listener;
        }
        
        /**
         * Gets the event priority
         * @return Event priority
         */
        public int getPriority() {
            return priority;
        }
        
        /**
         * Invokes the event method
         * @param event The event
         * @throws Exception If an error occurs
         */
        public void invoke(Event event) throws Exception {
            method.invoke(listener, event);
        }
    }
}

File: src/main/kotlin/net/lavaclient/client/events/EventManager.kt
/*
 * This file is part of LavaClient.
 * https://github.com/lavaclient/client
 */
package net.lavaclient.client.events

import net.lavaclient.client.LavaClient
import java.lang.reflect.Method

/**
 * Manages event subscriptions and posts events
 */
class EventManager {
    // Map of event classes to listeners
    private val listeners = mutableMapOf<Class<*>, MutableList<Pair<Any, Method>>>()
    
    /**
     * Register an object to receive events
     *
     * @param obj The object to register
     */
    fun register(obj: Any) {
        // Find all methods with Subscribe annotation
        for (method in obj.javaClass.declaredMethods) {
            if (method.isAnnotationPresent(Subscribe::class.java)) {
                // Method must have exactly one parameter
                if (method.parameterCount != 1) {
                    LavaClient.INSTANCE.warn("Event handler method ${method.name} in ${obj.javaClass.name} has invalid parameter count (expected 1, got ${method.parameterCount})")
                    continue
                }
                
                // Get the event class from the parameter
                val eventClass = method.parameterTypes[0]
                
                // Register the listener
                if (!listeners.containsKey(eventClass)) {
                    listeners[eventClass] = mutableListOf()
                }
                
                // Make the method accessible (in case it's private)
                method.isAccessible = true
                
                // Add the listener to the list
                listeners[eventClass]?.add(Pair(obj, method))
                
                LavaClient.INSTANCE.debug("Registered event handler ${method.name} in ${obj.javaClass.name} for event ${eventClass.name}")
            }
        }
    }
    
    /**
     * Unregister an object from receiving events
     *
     * @param obj The object to unregister
     */
    fun unregister(obj: Any) {
        // Remove all listeners for the object
        listeners.forEach { (_, listeners) ->
            listeners.removeIf { (listener, _) -> listener == obj }
        }
        
        // Remove empty lists
        listeners.entries.removeIf { (_, listeners) -> listeners.isEmpty() }
    }
    
    /**
     * Post an event to all listeners
     *
     * @param event The event to post
     * @return The event
     */
    fun <T> post(event: T): T {
        // Get the event class
        val eventClass = event!!.javaClass
        
        // Get all listeners for the event and its parent classes
        val eventListeners = mutableListOf<Pair<Any, Method>>()
        
        listeners.forEach { (clazz, listeners) ->
            if (clazz.isAssignableFrom(eventClass)) {
                eventListeners.addAll(listeners)
            }
        }
        
        // Call all listeners
        eventListeners.forEach { (obj, method) ->
            try {
                method.invoke(obj, event)
            } catch (e: Exception) {
                LavaClient.INSTANCE.error("Error calling event handler ${method.name} in ${obj.javaClass.name}")
                e.printStackTrace()
            }
            
            // Break if the event is cancelled
            if (event is Cancellable && event.isCancelled()) {
                return event
            }
        }
        
        return event
    }
}